### ISSUE TITLE
Create "User Guide" and/or "Examples" sections on RTD that correspond with workshop materials

### ISSUE THREAD

**Issue by joshuacwnewton** [spinalcordtoolbox/spinalcordtoolbox#2857]
> **Note:** This is a long-term suggestion for what our documentation could look like in the future. Feel free to give feedback!

### Current

Our [workshop PDF](https://www.icloud.com/keynote/0th8lcatyVPkM_W14zpjynr5g#SCT_Course_20200121) contains helpful mini-tutorials for how to use SCT in common processing pipelines. In some ways, the workshop PDF acts as an extension of our documentation, because we often ask people to check the latest SCT course (e.g. in [responses on spinalcordmri.org](https://forum.spinalcordmri.org/t/trouble-warping-to-template/460/2)). 

This is a bit tricky, though, because the PDF to the latest course is somewhat buried in our documentation (User Section -> Getting Started -> [Courses](https://spinalcordtoolbox.com/en/latest/user_section/getting-started.html#courses) -> iCloud link -> Download). Also, the PDF is quite a large document; it's hard to link to specific sections. So, users may become overwhelmed trying to skim through so much theory and material.

### Proposed

I think it could be good to have each of these mini-tutorials become pages on RTD as well. That way, we gain the ability to link to relevant guides when we need to explain how to use SCT. It would be more helpful for teaching/onboarding, too, because the material would be divided into easily-digestible chunks.

We could follow scikit-learn -- they maintain Sphinx-generated [User Guide](https://scikit-learn.org/stable/user_guide.html) and [Examples](https://scikit-learn.org/stable/auto_examples/index.html) pages with helpful guides for how to use each tool. As well, when a change is made to the code, contributors are instructed to update these pages at the same time so the documentation doesn't fall out of sync. 

If we adopt this approach, these pages could then become the foundation for future workshops. Because they would be kept up to date, it becomes much less of a headache to figure out what changes were made since the last workshop.

**Comment by jcohenadad** [comment]
terrific idea @joshuacwnewton !!! thank you so much for taking the lead on this epic task. I agree referring to the pdf is cumbersome

**Comment by joshuacwnewton** [comment]
Thanks for the positive words, @jcohenadad! My thinking, too, was that working on documentation could be a good way for me to get up to speed with the image processing/CV parts of SCT, which is where my interests lie! Hence this idea (and my comments on the other documentation issues).

**Comment by jcohenadad** [comment]
Another cool example of user-friendly handbook is the [Datalad handbook](http://handbook.datalad.org/en/latest/)

**Comment by jcohenadad** [comment]
possible help from https://pandoc.org/ to convert pdf into markdown for generating the html doc

**Comment by joshuacwnewton** [comment]
@jcohenadad After reflecting a bit, I'm having trouble understanding which of our problems (listed below) a datalad-style book would solve. Could you clarify what it is about the format that is appealing to you? It would be a big undertaking, so if our current RTD formatting is already sufficient, then it might be better to work on smaller, incremental changes to the RTD organization.

### Summary of issues with current documentation

1. **Documentation for CLI script usage is located in 3 separate places** (Argparse help descriptions, RTD tutorial pages, and SCT course material). Usage has to be written/updated in three different places.
     * Please note that this includes **CLI script usage in isolation**, as well as **CLI script usage as part of a pipeline**.
2. **RTD is missing information from SCT course slides.** The RTD pages by themselves do not fully explain SCT script usage, so they must be supplemented by the SCT course material.
3. **SCT course material is not discoverable.** The section is currently difficult to find on the RTD website.
3. **SCT course material is difficult to link to.** The large slide deck lacks static page/section links, and the course recordings are difficult to digest due to length.

**Comment by jcohenadad** [comment]
@joshuacwnewton thank you for putting together this nice summary. I agree with all of it. My idea behind the "book" was to have it completely integrated into RTD (not as a separate instance), so that already solves the "smaller, incremental changes" approach, which i agree 100% with.

Now, looking a bit more into it, I am realizing that having the integration of the book inside RTD might not be as trivial as I thought. So, we could, indeed, forget about the "book technology" and aim at transferring all the course material inside a well-structured section in RTD. 

Few considerations:
- As we add more images (and potentially animations/videos) inside the RTD, this will make the main SCT repos grow some fat, which we definitely want to avoid #1823. So, maybe we should think of separating the RTD content to another repos? ie: move it to the https://github.com/spinalcordtoolbox
- We need to keep in mind how this course material will be used: 
  - either by me (or other members of the team) during official SCT course: ie: website open, terminal open, and me going through the course material and copy/pasting the code listed in teh course material onto the terminal. So: a lot of back-and-forth between the two windows, and the transition should not be too confusing. But I don't see a problem with that (ie: will be similar to what we've done so far between Keynote/pdf and the Terminal).
  - either by users: sub-sections should be self-explanatory (segmentation, template, etc.), and at the same time users should be able to follow the whole course and think of it as a "whole" (ie: I would not want them to skip sections, etc., which a "book" presentation incite less to)

**Comment by joshuacwnewton** [comment]
> * As we add more images (and potentially animations/videos) inside the RTD, this will make the main SCT repos grow some fat, which we definitely want to avoid #1823. So, maybe we should think of separating the RTD content to another repos? ie: move it to the https://github.com/spinalcordtoolbox

My concern is that this would prevent us from being able to update the docs in the same PRs as the code changes. (If RTD was in a separate repo, we'd have to nag contributors to then open up another PR with another review process to update the docs alongside code changes.)

For images, `scikit-learn` does keep them in their project repo, as seen here: [`scikit-learn/doc/images`](https://github.com/scikit-learn/scikit-learn/tree/master/doc/images). Each of the images are quite small in size relative to the project as a whole. But, this assumes they're resized/compressed for display on the web. Since we have some very pretty high-res images from the course, the originals might need to be stored elsewhere as you mention. 

As for videos, there is always the `spinalcordtoolbox` YouTube channel, which we could embed using, say: https://github.com/readthedocs/readthedocs.org/issues/879

> either by me (or other members of the team) during official SCT course: ie: website open, terminal open, and me going through the course material and copy/pasting the code listed in teh course material onto the terminal. So: a lot of back-and-forth between the two windows, and the transition should not be too confusing. But I don't see a problem with that (ie: will be similar to what we've done so far between Keynote/pdf and the Terminal).

I'm glad you pointed this out. I think another part of this challenge is that the keynote has a dfferent writing style: 
* Slide decks have simplified bullet-points (as much of the explanation is done vocally by the speaker)
* Tutorials webpages are meant to stand alone, so they would have full paragraphs, etc.

So, these tutorial pages could end up giving a "wall of text" feeling (too much information to process) during a presentation. Perhaps this could be solved by having the headers to line up with the slide deck bullet points, so that the bigger text could make it easier to skim through. This is definitely something I want to keep in mind.

**Comment by jcohenadad** [comment]
> For images, one project we could look to as an example is `scikit-image`, which keeps their docs images in their project repo, as seen here: [`scikit-learn/doc/images`](https://github.com/scikit-learn/scikit-learn/tree/master/doc/images). Each of the images are quite small in size relative to the project as a whole. But, this assumes they're resized/compressed for display on the web. Since we have some very pretty high-res images from the course, the originals might need to be stored elsewhere as you mention.

hum, i'm not so sure about that. The "code" repos does have a `doc/` folder with some images, mostly institution logos, but the actual website is hosted on another repos, dedicated for the website, which includes a [much larger number of images](https://github.com/scikit-learn/scikit-learn.github.io/tree/master/0.23/_images). I haven't spent time figuring out why they also have a doc/ under the code repos though, but this configuration is what i had in mind for SCT. 

And videos, yes, i agree. They should definitely be embedded.

**Comment by joshuacwnewton** [comment]
> but the actual website is hosted on another repos, dedicated for the website, which includes a much larger number of images.

I believe the much larger number of images are matplotlib plots generated when the docs are built. :)

> I haven't spent time figuring out why they also have a doc/ under the code repos though, but this configuration is what i had in mind for SCT.

This quote from the [`scikit-learn.github.io`](https://github.com/scikit-learn/scikit-learn.github.io) might help: 

>  This repo is just hosting the HTML output of the documentation generated by the sphinx tool from the '.rst' text files of the scikit-learn documentation.

> The source of the documentation is managed in the doc folder of the main scikit-learn repo:

It's basically accomplishing the same thing as what we use ReadTheDocs for (building and hosting the docs), because they use GitHub Pages for hosting.

**Comment by jcohenadad** [comment]
Thank you for the clarification @joshuacwnewton . Generating figs on the fly is very cool but a bit difficult in our case (maybe some would be doable).

**Comment by joshuacwnewton** [comment]
As a quick update, I am about 75% done a new tutorial page for the Registration to template tutorial. I'm still sorting out how to structure the page, as the "Registration to template" section of the course is quite long (1hr24m of the 3hr course!). It might need to be broken up into multiple pages... 

Admittedly this is low-priority compared to more pressing issues (e.g. investigating user bugs) so I keep switching to higher-priority tasks. :sweat_smile:

**Comment by jcohenadad** [comment]
This is fantastic @joshuacwnewton i really appreciate your efforts in this.

**Comment by joshuacwnewton** [comment]
As an update, an upcoming SCT course has been tentatively scheduled for mid-November. So, we now have a more concrete deadline for finishing this. :blush: 

This issue is also an agenda item at the upcoming May 19th dev meeting.

**Comment by joshuacwnewton** [comment]
Small heads-up: I've started working on the remaining tutorials that are left after #3094. I figure that this is something useful to focus on whenever we have [a backlog of PRs that need review](https://user-images.githubusercontent.com/16181459/121901094-1ae16e00-ccf4-11eb-9ecb-ae6ee8a21612.png), since it doesn't create too much extra work for others. :sweat_smile:
